core/src/test/java/org/apache/iceberg/inmemory/InMemoryCatalog.java

```diff
@@ -0,0 +1,348 @@
+    }
+
+    return tableDb.keySet().stream()
+        .filter(t -> namespace.isEmpty() || t.namespace().equals(namespace))
+        .sorted(Comparator.comparing(TableIdentifier::toString))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public void renameTable(TableIdentifier fromTableIdentifier, TableIdentifier toTableIdentifier) {
+    if (fromTableIdentifier.equals(toTableIdentifier)) {
+      return;
+    }
+
+    if (!namespaceExists(toTableIdentifier.namespace())) {
+      throw new NoSuchNamespaceException(
+          "Cannot rename %s to %s because the namespace %s does not exist",
+          fromTableIdentifier, toTableIdentifier, toTableIdentifier.namespace());
+    }
+
+    if (!tableDb.containsKey(fromTableIdentifier)) {
+      throw new NoSuchTableException(
+          "Cannot rename %s to %s because the table %s does not exist",
+          fromTableIdentifier, toTableIdentifier, fromTableIdentifier);
+    }
+
+    if (tableDb.containsKey(toTableIdentifier)) {
+      throw new AlreadyExistsException(
+          "Cannot rename %s to %s: Table already exists", fromTableIdentifier, toTableIdentifier);
+    }
+
+    String fromLocation = tableDb.remove(fromTableIdentifier);
+    Preconditions.checkState(
+        null != fromLocation,
+        "Cannot rename from %s to %s because source table does not exist",
+        fromTableIdentifier,
+        toTableIdentifier);
+
+    tableDb.put(toTableIdentifier, fromLocation);
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace) {
+    createNamespace(namespace, Collections.emptyMap());
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    if (namespaceExists(namespace)) {
+      throw new AlreadyExistsException(
+          "Namespace already exists: %s. Cannot create namespace", namespace);
+    }
+
+    namespaceDb.put(namespace, ImmutableMap.copyOf(metadata));
+  }
+
+  @Override
+  public boolean namespaceExists(Namespace namespace) {
+    return namespaceDb.containsKey(namespace);
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) throws NamespaceNotEmptyException {
+    if (!namespaceExists(namespace)) {
+      return false;
+    }
+
+    List<TableIdentifier> tableIdentifiers = listTables(namespace);
+    if (!tableIdentifiers.isEmpty()) {
+      throw new NamespaceNotEmptyException(
+          "Namespace %s is not empty. Contains %d table(s).", namespace, tableIdentifiers.size());
+    }
+
+    return namespaceDb.remove(namespace) != null;
+  }
+
+  @Override
+  public boolean setProperties(Namespace namespace, Map<String, String> properties)
+      throws NoSuchNamespaceException {
+    if (!namespaceExists(namespace)) {
+      throw new NoSuchNamespaceException("Namespace does not exist: %s", namespace);
+    }
+
+    namespaceDb.computeIfPresent(
+        namespace,
+        (k, v) -> {
+          Map<String, String> newProperties = Maps.newHashMap(v);
```

Comment:
Should we use ImmutableMap.Builder just in case?

Reply:
makes sense, I've changed this to use `ImmutableMap.<String, String>builder().putAll(v).putAll(properties).buildKeepingLast()`